tools/config/__init__.py

```diff
@@ -836,10 +836,23 @@ def _generate_bootloader_build(self, rom_memories):
             yield Region("post_application", start, rom_end - start,
                          False, None)
         else:
+            last_address = rom_end
             if self.target.app_offset:
+                app_offset = integer(self.target.app_offset, 0)
+                last_address = app_offset
+
+                # Find the region to place the application so that
+                # both start (s) and end (e) of a section is beyond the offset
+                # provided. Then set s as the last possible address that the
+                # application can use.
+                for s, e in part.segments():
+                    if e > app_offset and s < app_offset:
+                        raise ConfigException("No available space to place the application in the given segments")
+                    if e > app_offset and s > app_offset:
+                        last_address = s
                 start = self._assign_new_offset(
-                    rom_start, start, self.target.app_offset, "application")
-            yield Region("application", start, rom_end - start,
+                    rom_start, start, self.target.app_offset, "application", True)
```

Comment:
This change will not protect users from the following scenario, which it was originally designed for:
Say you have the following:
 * A 16 K BL, end address is `0x4000`
 * Application configures it's offset to `0x3000`

This situation should result in an error. With this change, it might raise an exception from IntelHex which will not help the user. I would like to keep this deterministic error message.